src/API/Middlewares/RequestValidationMiddleware.ts

```diff
@@ -0,0 +1,27 @@
+import { IllegalArgumentException } from '@po.et/poet-js'
+import * as Joi from 'joi'
+interface RequestValidationParams {
+  readonly query?: (values: object) => {}
+  readonly body?: (values: object) => {}
+  readonly params?: (values: object) => {}
+  readonly options?: object
+}
+
+export const RequestValidationMiddleware = (requestValidationParams: RequestValidationParams) => async (
+  context: any,
+  next: () => Promise<any>
+) => {
+  try {
+    const { query, body, params, options } = requestValidationParams
+    /* tslint:disable:no-unused-expression */
+    body && (await Joi.validate(context.request.body, body(context.request.body), options))
+    query && (await Joi.validate(context.query, query(context.query), options))
+    params && (await Joi.validate(context.params, params(context.params), options))
+    return next()
+  } catch (exception) {
+    if (context.logger)
+      context.logger.error({ exception }, 'Error Caught at Middleware - Will Return Internal Server Error')
+
+    throw new IllegalArgumentException(exception)
```

Comment:
IllegalArgumentException takes an string as a parameter, not an object. If we're intentionally passing the exception object and expecting `.toString()` to be called on it, let's do that here to avoid room for interpretation.

We could also improve the `IllegalArgumentException` to support more precise parameters if needed.

Reply:
with the error message is enough. I can send the `exception .message`. 

Do you agree?

Reply:
Yeah, that sounds good.